src/highdicom/seg/sop.py

```diff
@@ -1328,35 +1387,50 @@ def from_dataset(
     def _build_ref_instance_lut(self) -> None:
         """Build lookup table for all instance referenced in the segmentation.
 
-        Builds a lookup table mapping the SOPInstanceUIDs of all datasets
-        referenced in the segmentation to a tuple containing the
-        StudyInstanceUID, SeriesInstanceUID and SOPInstanceUID.
+        Builds a table in the sqlite db containing all Study, Series and SOP
+        Instance UIDS for instances referenced in the segmentation.
 
         """
-        # Map sop uid to tuple (study uid, series uid, sop uid)
-        self._ref_ins_lut = OrderedDict()
+        instance_data = []
         if hasattr(self, 'ReferencedSeriesSequence'):
             for ref_series in self.ReferencedSeriesSequence:
                 for ref_ins in ref_series.ReferencedInstanceSequence:
-                    self._ref_ins_lut[ref_ins.ReferencedSOPInstanceUID] = (
-                        hd_UID(self.StudyInstanceUID),
-                        hd_UID(ref_series.SeriesInstanceUID),
-                        hd_UID(ref_ins.ReferencedSOPInstanceUID)
+                    instance_data.append(
+                        (
+                            self.StudyInstanceUID,
+                            ref_series.SeriesInstanceUID,
+                            ref_ins.ReferencedSOPInstanceUID
+                        )
                     )
         other_studies_kw = 'StudiesContainingOtherReferencedInstancesSequence'
         if hasattr(self, other_studies_kw):
             for ref_study in getattr(self, other_studies_kw):
                 for ref_series in ref_study.ReferencedSeriesSequence:
                     for ref_ins in ref_series.ReferencedInstanceSequence:
-                        self._ref_ins_lut[ref_ins.ReferencedSOPInstanceUID] = (
-                            hd_UID(ref_study.StudyInstanceUID),
-                            hd_UID(ref_series.SeriesInstanceUID),
-                            hd_UID(ref_ins.ReferencedSOPInstanceUID)
+                        instance_data.append(
+                            (
+                                ref_study.StudyInstanceUID,
+                                ref_series.SeriesInstanceUID,
+                                ref_ins.ReferencedSOPInstanceUID,
+                            )
                         )
 
-        self._source_sop_instance_uids = np.array(
-            list(self._ref_ins_lut.keys())
-        )
+        with self._db_con:
```

Comment:
I wonder whether it would make sense to abstract the database out of the `SOPClass` into a separate (private) class. That class could handle the database connection and expose methods to insert, find, and retrieve information. I feel that would be cleaner and would facilitate testing.

Reply:
Such a database service class could also become useful for managing frame-level data access for other multi-frame SOP class instances in the future (Parametric Map, etc.).

Reply:
As of [8f5172f](https://github.com/ImagingDataCommons/highdicom/pull/208/commits/8f5172f2f692937eee71574ab81c4bc95e32d87e) the database related logic has been significantly refactored into a new `_SegDBManager` class